Flipr integration

cnico opened the pull request:
## Proposed change
Addition of a new integration named flipr.
This integration is for a pool device that measures pool datas like ph, chlorine and temperature. It simply poll them from the cloud provider and expose them as entities in homeassistant.
See [https://goflipr.com/](https://goflipr.com/) for the website of the manufactor.


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
No configuration in yaml files: this integration is configured through the UI only.
The API used to poll the cloud service is in the flipr-api module I pushed on pypi.

- This PR fixes or closes issue: None
- This PR is related to issue: None
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16596
- Link to brand pull request: https://github.com/home-assistant/brands/pull/2261

## Checklist

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

I don't know what quality scale my flipr integration could have reached. I did my best for it.


To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cnico committed:
Réalisation du config_flow pour l'installation

cnico committed:
Fonctionnement OK des sensors flipr

cnico committed:
Fonctionnement OK des sensors flipr

cnico committed:
Revert addition of a print for debug

cnico committed:
Revert addition of a print for debug

cnico committed:
report modif after rebase

Commit:
Fix suite au passage des linters

Commit:
ajout de deps flipr dans requirements_*.txt

Commit:
passing test

cnico committed:
Ajout du cryptage du mot de passe dans la config

cnico committed:
Gestion des cas 0 ou plusieurs flipr dans le config_flow

Commit:
Test correction with encrypted password

cnico committed:
add unit test on sensors

cnico committed:
comment correction

cnico committed:
isort correction

cnico committed:
Upgrade flipr-api dependency to 1.2.0

cnico committed:
Improve of tests coverage

frenck commented on homeassistant/components/flipr/translations/fr.json:
```diff
@@ -0,0 +1,30 @@
+{
```
Please do not add additional translations in the initial PR. Translations can be managed via lokalise after this PR has been merged.

cnico replied:
OK, correction done via a new commit and push.

frenck commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,86 @@
+"""Sensor platform for the Flipr's pool_sensor."""
```
This PR contains 2 platforms: binary_sensor & sensor.
Please limit the initial PR to add an integration to a single platform.

cnico replied:
It would require some rework to remove the binary_sensor (which is less important) that I would have to put in another PR right after this one and It would imply to retest everything.
Is it a blocker one or could you accept it as is ?

frenck replied:
Yes, this is a blocker and is described in our developer documentation:

<https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible>

cnico replied:
Done

frenck commented on homeassistant/components/flipr/manifest.json:
```diff
@@ -0,0 +1,16 @@
+  "requirements": [
+    "flipr-api==1.2.0",
+    "pycryptodome==3.10.1"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
These are empty defaults that can be removed.

cnico replied:
OK, correction done via a new commit and push.

frenck commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,86 @@
+    "ph": {"unit": "ph", "icon": "mdi:pool", "name": "PH", "device_class": None},
+    "temperature": {
+        "unit": TEMP_CELSIUS,
+        "icon": "mdi:coolant-temperature",
+        "name": "Water Temp",
+        "device_class": DEVICE_CLASS_TEMPERATURE,
```
Either set a device class or icon, not both.

cnico replied:
OK, correction done via a new commit and push.

frenck reviewed (commented):
Some quick small first comments to get something going 👍

cnico committed:
Corrections for PR review

cnico committed:
correction of sensor boolean logic and unit tests

cnico commented:
@frenck : thanks for your review. I hope the corrections I have done are correct for you.

cnico committed:
Update flipr api for a small bug correction

cnico commented:
Hi @frenck,

It has been a long time since I created this pull request.
Is it still planned / in progress to be integrated or is it abandonned ?

I am not used to homeassistant's project delays so I find it quite long but I understand also there is a lot of activity to manage on the project.

Regards,

cnico committed:
Update flipr api for a small bug correction

cnico committed:
Removal of binary_sensor as requested for merge request

cnico committed:
Addition of iot-class in manifest

cnico commented:
As your CI controls evolved since my first proposal, I had to do a rebase and add the iot_class manifest information.
Now, that there is only one sensor, can you validate this merge request ?

iMicknl commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,86 @@
+    @property
+    def device_state_attributes(self):
+        """Return device attributes."""
+        return {ATTR_ATTRIBUTION: ATTRIBUTION}
```
Perhaps you want to implement `device_info` as well? See https://developers.home-assistant.io/docs/device_registry_index/#defining-devices.

This will allow you to group all entities from a particular flipr_id under one device in Home Assistant.


```python3
    @property
    def device_info(self):
        return {
            "identifiers": {
                # Serial numbers are unique identifiers within a specific domain
                (DOMAIN, self.flipr_id)
            },
            "name": self.name, #device name
            "manufacturer": self.light.manufacturername,
            "model": self.light.productname, # optional
            "sw_version": self.light.swversion, # optional
        }
```

cnico replied:
OK, I added it.

iMicknl reviewed (commented):
Hopefully my comments are helpful. Great to see so many tests as well, nice contribution!

cnico commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,119 @@
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: Dict) -> bool:
```
OK, I removed this method.

cnico commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,119 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Flipr from a config entry."""
+    _LOGGER.debug("async_setup_entry starting")
+
```
Yep, I integrated this change in my commit.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+            except HTTPError:
+                errors = {}
+                errors["base"] = "invalid_auth"
+                return self._show_setup_form(errors)
```
Actually, I do not know how the server behave in bad situations since I have only see it up and running.
I added your proposal in the code to catch other errors to improve it.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+                self._possible_flipr_ids = flipr_ids
+                return await self.async_step_flipr_id()
+
+        # No check for data retrieval here.
```
It is a comment I forgot to remove from the integration I took examples of code.
I removed this comment line.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+        # No check for data retrieval here.
+
+        # Check if already configured
+        await self.async_set_unique_id("FLIPR_with_ID_" + self._flipr_id)
```
The prefix was to be sure there could not be another integration with the same ID if we have no luck.
For information, a flipr_id is just 5 characters long.
Perhaps that the unique id is relative to the current DOMAIN so I agree there is no probability of collision in such a case.
I removed the prefix as suggested.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+        self._abort_if_unique_id_configured()
+
+        # Encrypt password before storing it in the config json file.
+        crypted_password = encrypt_data(self._password, self._flipr_id)
```
Actually, when, for debug reasons, I searched for my configuration information in the data files, I fall on my password visible without being encrypted. I know my solution is not perfect and a clever attacker that would track the code source of the integration would be able to decypher it but it seems better for me never to store a password unencrypted.
I will start a discussion as you proposed.

cnico replied:
Discussion opened here : https://github.com/home-assistant/architecture/discussions/556

cnico replied:
After, re-thinking of it, I removed all my code related to password encryption in the flipr integration.
I think the discussion and feature will continue in the related discussion.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+        crypted_password = encrypt_data(self._password, self._flipr_id)
+
+        return self.async_create_entry(
+            title="Flipr device - " + self._flipr_id,
```
OK, I integrated this change in my commit.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+        if not self._flipr_id:
+            try:
+                flipr_ids = await self._authenticate_and_search_flipr()
+            except HTTPError:
```
Actually, I don't know as I am not a developer of the API and it is not clearly documented.
I did not find a better error handling to catch invalid auth more precisely.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+                        }
+                    ),
+                )
+            # Else : no flipr_id found. Allow the user to enter it.
```
In my experience, it is a 5 chars length string but I can not guarantee it. In the API documentation, it is simply defined as a string.

cnico commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,128 @@
+
+    async def async_step_import(self, user_input):
+        """Import a config entry."""
+        return await self.async_step_user(user_input)
```
OK, done.

cnico commented on homeassistant/components/flipr/const.py:
```diff
@@ -0,0 +1,9 @@
+
+DOMAIN = "flipr"
+
+CONF_USERNAME = "email"
```
OK, I replaced my constants with homeassistant.const ones.
The username is an email so I kept the variables with name username and I used the const : CONF_EMAIL.
It so has no impact on strings.json and translations.

cnico commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,86 @@
+        "name": "Chlorine",
+        "device_class": None,
+    },
+    "ph": {"unit": "ph", "icon": "mdi:pool", "name": "PH", "device_class": None},
```
OK, changed it.

cnico commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,86 @@
+    "date_time": {
+        "unit": None,
+        "icon": None,
+        "name": "Date Measure",
```
Yes, I changed it.

cnico committed:
Improvements following PR review

cnico committed:
api small bug correction with logged data

cnico committed:
Password encryption in config removal : this should be transparent and in core

cnico commented:
Thank you @iMicknl for your review and comments. I integrated all your proposals and give some comments for each. I pushed the new version of the code in the branch after having retested everything works fine and it's OK for me.
I let you poursue the review and pull request process.

cnico commented:
I resolved the conversation and don't know if it was right or not to do it.
I don't understand why the CI builds did not re-run after my push... I thought it was because of the open review conversations... but apparently not.

cnico committed:
removal of pycriptdome dependency which is no more usefull

iMicknl commented:
> I don't understand why the CI builds did not re-run after my push... I thought it was because of the open review conversations... but apparently not.

Unfortunately this is [a change on the GitHub side](https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/), you will need to wait until a maintainer approves the workflow...

iMicknl commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,109 @@
+
+    coordinator = FliprDataUpdateCoordinator(hass, entry)
+
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

and you should be able to remove the check for `coordinator.last_update_success`

cnico replied:
OK, I did not know this method.
This will raise me to 100% test coverage :)

iMicknl commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,109 @@
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
```
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)

    return True
```

cnico replied:
OK, took this removal of code.

iMicknl commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,109 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
```
```suggestion
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

cnico replied:
OK, took this removal of code.

iMicknl commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,98 @@
+
+    @property
+    def device_info(self):
+        """Define device information global to entities."""
```
Would it make sense to move this logic to the `FliprEntity`?

cnico replied:
Yes, your are right : I moved it to FliprEntity

iMicknl commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+            # It may cause no data found afterwards in metrics...
+            return self.async_show_form(
+                step_id="flipr_id",
+                data_schema=vol.Schema({vol.Required(CONF_FLIPR_ID): str}),
```
It seems that flipr id's always need to be 9 characters? Would it make sense to validate this? Or even better, are you able to validate via the API if a flipr id is correct and if it will return data?

I don't know enough about the device / API to give a good advice, but the method seem very error prone. Does it happen often that no flipr id's are returned when you are logged in?

cnico replied:
No, there is no rule on flipr id size from my knowledge.
And there is no API to valide a flipr_id.
I imagined this in the case the user would use homeassistant's flipr integration and he has not initialized or configured it  through the official mobile app. This should be a very rare case so after rethinking of it, it is certainly better to remove this case from the config setup and explain it through an error message.
This is what I did : I changed it to an error case during setup.

iMicknl replied:
It would make sense to make the Config Flow as simple as possible. If someone has multiple Flipr devices linked to one account, do you need to create a config entry for every device (by Flipr Id)?

Best would be to just pull all the linked Flipr ID's on load / reload and create devices + entities based on the available devices.

iMicknl replied:
@cnico wouldn't it make sense to remove the logic where you only select a single Flipr ID? Doesn't it make sense to do this in __init__ thus that all devices linked to your account will be added? Are there any usecases where you only want to select a single flipr id from your account?

```client.search_flipr_ids```

cnico replied:
Yes, you are right, if the user has many fliprs, which is not my case for now, the best thing is probably to directly add all entities. If some are useless, it should not be a problem.
I will refactor that way and push it.

iMicknl commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,98 @@
+    @property
+    def name(self):
+        """Return the name of the particular component."""
+        return f"Flipr {self.flipr_id} {SENSORS[self.info_type]['name']}"
```
Would be good to have a core maintainer comment on this one. Since we have the device registry, different physical devices are already divided by Flipr id, thus do we still need the name of the integration and name of the device in every entity.

I can imagine that this is most readable one, but I wonder what the current guidance is. (also for other integrations).

cnico replied:
Yes I prefer to have the "flipr" prefix because otherwise, when you configure lovelace content, it is quite harder to find the entities...
I checked my global entity list and found that for my samsung phones, all entities are prefixed by this "sm_[model_number]".

iMicknl replied:
>For name and entity_id I don't think we have a standard in that regard. I usually go with what the contributor suggests as name.

Let's keep your suggestions ;-).

cnico replied:
OK, cool.

iMicknl reviewed (commented):
Changes look good! Good that you took out the encryption part, and moved it to an architecture discussion.

I had some minor comments from my side, and an open question for core maintainers.

cnico committed:
Corrections following review

cnico commented:
@iMicknl , I integrated your review content, thanks again.
For the question to the core team, I let you see. I tell you my preference but if the rule is to remove such prefix, I will remove it.

cnico committed:
Replace thee one config per flipr implementation by a one config for many fliprs

cnico commented:
@iMicknl, it was not easy for me but I did it : now, there is one config for many fliprs. 
It works with my situation of one flipr and all tests and checks are good for me.
I hope, it is good now to finalize the pull request.
Once it will be done, there will be another PR with the binary_sensor for flipr alerts that I removed as required by frenck.

cnico commented:
Hi,
@frenck , @iMicknl , what is the next step for this PR ?

cnico,

tmenguy commented:
Yep HA reviewers … when will we enjoy our flipr in HA … everything seems fine now

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,111 @@
+    for flipr_id in flipr_ids:
+        coordinator = FliprDataUpdateCoordinator(hass, entry, flipr_id)
+        await coordinator.async_config_entry_first_refresh()
+        hass.data[DOMAIN][entry.entry_id + flipr_id] = coordinator
```
I think this is cleaner, also on unload you can just pop de entry ID without iterating again.
```suggestion
        hass.data[DOMAIN][entry.entry_id][flipr_id] = coordinator
```

cnico replied:
I integrated it but it did not work at runtime. I had to had this line before adding coordinators : 
hass.data[DOMAIN][entry.entry_id] = {}

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,111 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
+        flipr_ids = entry.data[CONF_FLIPR_IDS].split(",")
+        for flipr_id in flipr_ids:
+            hass.data[DOMAIN].pop(entry.entry_id + flipr_id)
```
```suggestion
        hass.data[DOMAIN].pop(entry.entry_id)
```

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,111 @@
+
+        # Establishes the connection.
+        self.client = FliprAPIRestClient(username, password)
+        self.hass = hass
```
This is already in `DataUpdateCoordinator`

```suggestion
```

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,111 @@
+    def __init__(self, coordinator, flipr_id, info_type):
+        """Initialize Flipr sensor."""
+        super().__init__(coordinator)
+        self._unique_id = f"{flipr_id}-{info_type}"
```
Saves a little bit of code: <https://developers.home-assistant.io/docs/core/entity#entity-class-attributes>
```suggestion
        self._attr_unique_id = f"{flipr_id}-{info_type}"
        self._attr_device_info = {
            "identifiers": {(DOMAIN, flipr_id)},
            "name": NAME,
            "manufacturer": MANUFACTURER,
        }
```

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,111 @@
+        self._unique_id = f"{flipr_id}-{info_type}"
+        self.info_type = info_type
+        self.flipr_id = flipr_id
+
+    @property
+    def unique_id(self):
+        """Return a unique id."""
+        return self._unique_id
+
+    @property
+    def device_info(self):
+        """Define device information global to entities."""
+        return {
+            "identifiers": {
+                # Serial numbers are unique identifiers within a specific domain
+                (DOMAIN, self.flipr_id)
+            },
+            "name": NAME,
+            "manufacturer": MANUFACTURER,
+        }
```
```suggestion
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+
+from .const import CONF_FLIPR_IDS
+from .const import DOMAIN  # pylint:disable=unused-import
```
```suggestion
from .const import DOMAIN
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    """Handle a config flow for Flipr."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
This isn't used anymore

```suggestion
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize Flipr config flow."""
+        self._username = None
+        self._password = None
+        self._flipr_ids = None
```
```suggestion
    _username: str | None = None
    _password: str | None = None
    _flipr_ids: str | None = None
```

cnico replied:
This does not work as is since python 3.10 is not the current version (at least in the devcontainer of the upstream dev branch). So I used the old style with typing.Union for now.

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        self._username = user_input[CONF_EMAIL]
+        self._password = user_input[CONF_PASSWORD]
+
+        if not self._flipr_ids:
```
```suggestion
        errors = {}
        if not self._flipr_ids:
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+            try:
+                flipr_ids = await self._authenticate_and_search_flipr()
+            except HTTPError:
+                errors = {}
+                errors["base"] = "invalid_auth"
+                return self._show_setup_form(errors)
```
```suggestion
                errors["base"] = "invalid_auth"
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+                errors["base"] = "invalid_auth"
+                return self._show_setup_form(errors)
+            except (Timeout, ConnectionError):
+                errors = {}
+                errors["base"] = "cannot_connect"
+                return self._show_setup_form(errors)
```
```suggestion
                errors["base"] = "cannot_connect"
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+                errors["base"] = "cannot_connect"
+                return self._show_setup_form(errors)
+            except Exception as exception:  # pylint: disable=broad-except
+                errors = {}
+                errors["base"] = "unknown"
+                _LOGGER.exception(exception)
+                return self._show_setup_form(errors)
```
```suggestion
                errors["base"] = "unknown"
                _LOGGER.exception(exception)
```

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+                _LOGGER.exception(exception)
+                return self._show_setup_form(errors)
+
+            if len(flipr_ids) == 0:
+                # No flipr_id found. Tell the user with an error message.
+                errors = {}
+                errors["base"] = "no_flipr_id_found"
+                return self._show_setup_form(errors)
```
```suggestion
            if len(flipr_ids) == 0:
                # No flipr_id found. Tell the user with an error message.
                errors["base"] = "no_flipr_id_found"

            if errors:
                return self._show_setup_form(errors)
```

cnico replied:
To make it work, at runtime, I added a check on errors before len(flipr_ids) because otherwise, flipr_ids could be not assigned and it crashes.

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+                errors["base"] = "no_flipr_id_found"
+                return self._show_setup_form(errors)
+
+            # If multiple flipr found (rare case), we concatene the ids to create multiple devices in this configuration.
```
```suggestion
            # If multiple flipr found (rare case), we concatenate the ids to create multiple devices in this configuration.
```

Additionally, why not store a list instead of concatinating?

cnico replied:
I considered it and managed to change it to store a list.

frenck commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,95 @@
+            self._flipr_ids = ",".join(flipr_ids)
+
+        # Check if already configured
+        await self.async_set_unique_id(self._flipr_ids)
```
Should the config flow ask which flipr to add? So the right ID can be set and added to HA?

cnico replied:
I think, someone who has many flipr will prefer to have all of them configured in one shot instead of having to configure them each manually by selecting the proposed flipr_id.
This case of many flipr should be very rare since when a flipr is not working, it is removed by the manufacturer from the customer account, and not so many people have two or more swimming pools needing more than one flipr.

cnico committed:
frenck review : cleaner

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : cleaner

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : useless

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : cleaner

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : useless code

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better error handling

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : remove pylint ignore

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better error handling

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better error handling

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : orthographe

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : useless code

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better code writing

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better error handling

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
frenck review : better error handling

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
Unit test corrections and runtime corrections following frenck code review

cnico commented:
Hi @frenck, 
I integrated all of your points. I did some tests and a few corrections to make it work.
For me it is functional with all unit tests.
I also did a rebase on the upstream dev branch so it is tested with up to date code of homeassistant core.
I let you advance the PR workflow.

frenck commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,97 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Flipr from a config entry."""
+    _LOGGER.debug("async_setup_entry starting")
```
Home Assistant already logs this.

```suggestion
```

cnico committed:
Delete useless log

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico committed:
syntax correction

cnico committed:
config flow in 2 steps with a select of flipr_id if multiple are available

cnico committed:
Comment correction

cnico commented:
Hi @frenck 
I corrected the points you reviewed.
After reading many times your last point on unique_id and removal problem, I finally understood that the problem is related to using a list of flipr ids in one config flow. So I fall back to my previous implementation that manages it by asking the user to choose the correct flipr_id in a select box through a second step when the user has more than one flipr (rare case).
I also corrected the problem detected by the CI bot. It should pass at next run.
Everything works fine from my point of view.

frenck commented on homeassistant/components/flipr/strings.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "title": "Flipr",
```
This doesn't need translating

```suggestion
```

frenck commented on homeassistant/components/flipr/translations/en.json:
```diff
@@ -0,0 +1,31 @@
+                "title": "Flipr device"
+            }
+        }
+    },
+    "title": "Flipr"
```
```suggestion
    }
```

frenck reviewed (changes requested):
Besides some newer patterns we could apply to reduce amount of code a bit, I think this looks ok 👍 

One small translation suggestion that we should address before merging.

cnico committed:
Apply suggestions from code review

title removal in json's strings

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico commented:
OK, your suggestion of strings removal has been pushed. 
You can finalize the merge :)

cnico reviewed (commented):
All changes reviewed.

frenck commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,85 @@
+    @property
+    def state(self):
+        """State of the sensor."""
+        return self.coordinator.data[self.info_type]
```
Just noticed, based on the tests, this can be a `datetime` object as well, right??
If so, we need to make sure it is returned as an iso valid timestamp or else it will have issues on the frontend in certain browsers (e.g., Safari). Besides, `datetime` isn't a valid state.

```suggestion
        state = self.coordinator.data[self.info_type]
        if isinstance(state, datetime):
            return state.isoformat()
        return state
```

cnico replied:
yes, that's right, it can be a datetime.

frenck commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,85 @@
+"""Sensor platform for the Flipr's pool_sensor."""
+from homeassistant.const import (
```
```suggestion
from datetime import datetime

from homeassistant.const import (
```

frenck commented on tests/components/flipr/test_sensors.py:
```diff
@@ -0,0 +1,92 @@
+    assert state
+    assert state.attributes.get(ATTR_ICON) is None
+    assert state.attributes.get(ATTR_UNIT_OF_MEASUREMENT) is None
+    assert state.state == "2021-02-15 09:10:32+00:00"
```
```suggestion
    assert state.state == "2021-02-15T09:10:32+00:00"
```

cnico committed:
Correct support of datetime display for the sensor

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

cnico reviewed (commented):
OK for me. Another review and tests done.

frenck reviewed (approved):
LGTM @cnico 👍 🥇 

Thanks! 🎉

MartinHjelmare commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,90 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
```
Missing return value typing. It should return a boolean. Please type the whole signature when adding type annotations to a function.

MartinHjelmare commented on homeassistant/components/flipr/__init__.py:
```diff
@@ -0,0 +1,90 @@
+        password = entry.data[CONF_PASSWORD]
+        self.flipr_id = entry.data[CONF_FLIPR_ID]
+
+        _LOGGER.debug("Config entry values : %s, %s", username, self.flipr_id)
```
We don't want side effects in init methods. Please move or remove the logging.

MartinHjelmare commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+                errors["base"] = "unknown"
+                _LOGGER.exception(exception)
+
+            if not errors and len(flipr_ids) == 0:
```
```py
if not errors and not flipr_ids:
```

MartinHjelmare commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+            FliprAPIRestClient, self._username, self._password
+        )
+
+        flipr_ids = await self.hass.async_add_executor_job(client.search_flipr_ids)
```
It's more efficient to create a function that does both sync calls and then schedule that function once on the executor. It's expensive to switch context.

MartinHjelmare commented on homeassistant/components/flipr/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+    async def _authenticate_and_search_flipr(self) -> list[str]:
+        """Validate the username and password provided and searches for a flipr id."""
+        client = await self.hass.async_add_executor_job(
+            FliprAPIRestClient, self._username, self._password
```
Does this have side effects? We don't schedule this instantiation on the executor when setting up the integration.

MartinHjelmare commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,90 @@
+
+SENSORS = {
+    "chlorine": {
+        "unit": "mV",
```
Please use as many common constants as possible.

https://github.com/home-assistant/core/blob/8d7a136fc4e90b4f0136fc410c447598eea290cd/homeassistant/const.py#L416

MartinHjelmare commented on homeassistant/components/flipr/sensor.py:
```diff
@@ -0,0 +1,90 @@
+    async_add_entities(sensors_list, True)
+
+
+class FliprSensor(FliprEntity, Entity):
```
Please inherit `SensorEntity` instead of `Entity`.

MartinHjelmare commented on tests/components/flipr/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+async def test_invalid_credential(hass, mock_setup):
+    """Test invalid credential."""
+    with patch(
+        "flipr_api.FliprAPIRestClient.search_flipr_ids", side_effect=HTTPError()
```
Does this patch in the right place?

https://docs.python.org/3/library/unittest.mock.html#where-to-patch

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
